connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -1320,89 +1338,188 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
-                                                "haven't stopped completely yet or the connector was resumed before the request to alter its offsets could be successfully " +
-                                                "completed. If the connector is in a stopped state, this operation can be safely retried. If it doesn't eventually succeed, the " +
-                                                "Connect cluster may need to be restarted to get rid of the zombie sink tasks."),
-                                        null);
-                            } else {
-                                cb.onCompletion(new ConnectException("Failed to alter consumer group offsets for connector " + connName, error), null);
-                            }
-                        } else {
-                            completeAlterOffsetsCallback(alterOffsetsResult, cb);
-                        }
-                    }).whenComplete((ignored, ignoredError) -> {
-                        // errors originating from the original future are handled in the prior whenComplete invocation which isn't expected to throw
-                        // an exception itself, and we can thus ignore the error here
-                        Utils.closeQuietly(admin, "Offset alter admin for sink connector " + connName);
-                    });
+                    if (isReset) {
+                        resetSinkConnectorOffsets(connName, groupId, admin, cb, alterOffsetsResult);
+                    } else {
+                        alterSinkConnectorOffsets(connName, groupId, admin, offsetsToWrite, cb, alterOffsetsResult);
+                    }
                 } catch (Throwable t) {
-                    Utils.closeQuietly(admin, "Offset alter admin for sink connector " + connName);
+                    Utils.closeQuietly(admin, "Offset modification admin for sink connector " + connName);
                     throw t;
                 }
             } catch (Throwable t) {
-                cb.onCompletion(ConnectUtils.maybeWrap(t, "Failed to alter offsets for sink connector " + connName), null);
+                cb.onCompletion(ConnectUtils.maybeWrap(t, "Failed to modify offsets for sink connector " + connName), null);
             }
         }));
     }
 
     /**
-     * Alter a source connector's offsets.
+     * Alter a sink connector's consumer group offsets. This is done via calls to {@link Admin#alterConsumerGroupOffsets}
+     * and / or {@link Admin#deleteConsumerGroupOffsets}.
      *
-     * @param connName the name of the source connector whose offsets are to be altered
+     * @param connName the name of the sink connector whose offsets are to be altered
+     * @param groupId the sink connector's consumer group ID
+     * @param admin the {@link Admin admin client} to be used for altering the consumer group offsets; should be closed after use
```

Comment:
We close the admin client in this method, right? I think we can be clearer about that:
```suggestion
     * @param admin the {@link Admin admin client} to be used for altering the consumer group offsets; will be closed after use
```

Reply:
Good catch, thanks. I think this was leftover from a previous refactor (earlier it was the caller's responsibility to close the admin client).